File: tests/authoring/Applicability/AppliesToDirective.fs
// Licensed to Elasticsearch B.V under one or more agreements.
// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information

module ``product availability``.``yaml directive``

open Elastic.Markdown.Myst.Directives.AppliesTo
open Elastic.Markdown.Myst.FrontMatter
open authoring
open authoring.MarkdownDocumentAssertions
open Swensen.Unquote
open Xunit

type ``piggy back off yaml formatting`` () =
    static let markdown = Setup.Markdown """
```yaml {applies_to}
serverless:
  security: ga 9.0.0
  elasticsearch: beta 9.1.0
  observability: removed 9.2.0
```
"""

    [<Fact>]
    let ``parses to AppliesDirective`` () =
        let directives = markdown |> converts "index.md" |> parses<AppliesToDirective>
        test <@ directives.Length = 1 @>

        directives |> appliesToDirective (ApplicableTo(
            Serverless=ServerlessProjectApplicability(
                Security=AppliesCollection.op_Explicit "ga 9.0.0",
                Elasticsearch=AppliesCollection.op_Explicit "beta 9.1.0",
                Observability=AppliesCollection.op_Explicit "removed 9.2.0"
            )
        ))

type ``plain block`` () =
    static let markdown = Setup.Markdown """
```{applies_to}
serverless:
  security: ga 9.0.0
  elasticsearch: beta 9.1.0
  observability: removed 9.2.0
apm_agent_dotnet: ga 9.0
apm_agent_node: ga 10.0
```
"""

    [<Fact>]
    let ``parses to AppliesDirective`` () =
        let directives = markdown |> converts "index.md" |> parses<AppliesToDirective>
        test <@ directives.Length = 1 @>

        directives |> appliesToDirective (ApplicableTo(
            Serverless=ServerlessProjectApplicability(
                Security=AppliesCollection.op_Explicit "ga 9.0.0",
                Elasticsearch=AppliesCollection.op_Explicit "beta 9.1.0",
                Observability=AppliesCollection.op_Explicit "removed 9.2.0"
            ),
            ProductApplicability=ProductApplicability(
                ApmAgentDotnet=AppliesCollection.op_Explicit "ga 9.0",
                ApmAgentNode=AppliesCollection.op_Explicit "ga 10.0"
            )
        ))

type ``warns on old syntax`` () =
    static let markdown = Setup.Markdown """
```{applies_to}
:hosted: all
```
"""
    [<Fact>]
    let ``has no errors`` () = markdown |> hasNoErrors

    [<Fact>]
    let ``warns on bad syntax`` () =
        markdown |> hasWarning "Applies block does not use valid yaml keys: :hosted"

type ``warns on invalid keys`` () =
    static let markdown = Setup.Markdown """
```{applies_to}
hosted: all
```
"""
    [<Fact>]
    let ``has no errors`` () = markdown |> hasNoErrors

    [<Fact>]
    let ``warns on bad syntax`` () =
        markdown |> hasWarning "Applies block does not support the following keys: hosted"

File: tests/Elastic.Authoring.Tests/Applicability/AppliesToDirective.cs
// Licensed to Elasticsearch B.V under one or more agreements.
// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information

using AwesomeAssertions;
using Elastic.Documentation.AppliesTo;

// Alias needed: file namespace contains 'AppliesToDirective' as a segment, which shadows the type.
using AppliesToDirectiveType = Elastic.Markdown.Myst.Directives.AppliesTo.AppliesToDirective;

namespace Elastic.Authoring.Tests.Applicability;

public class PiggyBackOffYamlFormatting : MarkdownTest
{
	protected override string Markdown =>
		"""
		```yaml {applies_to}
		serverless:
		  security: ga
		  elasticsearch: beta
		  observability: removed
		```
		""";

	[Test, DisplayName("parses to AppliesDirective")]
	public async Task ParsesAppliesDirective()
	{
		var parsesTask = Docs.Converts("index.md").Parses<AppliesToDirectiveType>();
		(await parsesTask).Should().HaveCount(1);
		await parsesTask.AppliesToDirective(new ApplicableTo
		{
			Serverless = new ServerlessProjectApplicability
			{
				Security = Applies("ga"),
				Elasticsearch = Applies("beta"),
				Observability = Applies("removed"),
			},
		});
	}
}

public class PlainBlock : MarkdownTest
{
	protected override string Markdown =>
		"""
		```{applies_to}
		serverless:
		  security: ga
		  elasticsearch: beta
		  observability: removed
		apm_agent_dotnet: ga 9.0
		apm_agent_node: ga 10.0
		```
		""";

	[Test, DisplayName("parses to AppliesDirective")]
	public async Task ParsesAppliesDirective()
	{
		var parsesTask = Docs.Converts("index.md").Parses<AppliesToDirectiveType>();
		(await parsesTask).Should().HaveCount(1);
		await parsesTask.AppliesToDirective(new ApplicableTo
		{
			Serverless = new ServerlessProjectApplicability
			{
				Security = Applies("ga"),
				Elasticsearch = Applies("beta"),
				Observability = Applies("removed"),
			},
			ProductApplicability = new ProductApplicability { ApmAgentDotnet = Applies("ga 9.0"), ApmAgentNode = Applies("ga 10.0"), },
		});
	}
}

public class WarnsOnOldSyntax : MarkdownTest
{
	protected override string Markdown => """
		```{applies_to}
		:hosted: all
		```
		""";

	[Test, DisplayName("has no errors")]
	public async Task HasNoErrors() => await Docs.HasNoErrors();

	[Test, DisplayName("warns on bad syntax")]
	public async Task WarnsOnBadSyntax() => await Docs.HasWarning("Applies block does not use valid yaml keys: :hosted");
}

public class WarnsOnInvalidKeys : MarkdownTest
{
	protected override string Markdown => """
		```{applies_to}
		hosted: all
		```
		""";

	[Test, DisplayName("has no errors")]
	public async Task HasNoErrors() => await Docs.HasNoErrors();

	[Test, DisplayName("warns on bad syntax")]
	public async Task WarnsOnBadSyntax() => await Docs.HasWarning("Applies block does not support the following keys: hosted");
}
